src/path_object_gui.cpp

```diff
@@ -0,0 +1,203 @@
+#include "viewport.h"
+#include "window.h"
+
+/**
+ * %PathObject build GUI.
+ * @ingroup gui_group
+ */
+class PathObjectGui : public GuiWindow {
+public:
+	PathObjectGui();
+	~PathObjectGui();
+
+	void OnClick(WidgetNumber wid, const Point16 &pos) override;
+	void SetWidgetStringParameters(WidgetNumber wid_num) const override;
+
+	void SelectorMouseMoveEvent(Viewport *vp, const Point16 &pos) override;
+	void SelectorMouseButtonEvent(uint8 state) override;
+
+	void SetType(const PathObjectType *t);
+
+private:
+	RideMouseMode path_object_sel;               ///< Mouse selector for building path objects.
+	const PathObjectType *type;                  ///< Type currently being placed (or \c nullptr).
+	std::unique_ptr<PathObjectInstance> object;  ///< Item being placed.
+};
+
+/**
+ * Widget numbers of the PathObject build GUI.
+ * @ingroup gui_group
+ */
+enum PathObjectWidgets {
+	POBJ_BUTTON_BENCH,  ///< Place bench button.
+	POBJ_BUTTON_BIN,    ///< Place bin   button.
+	POBJ_BUTTON_LAMP,   ///< Place lamp  button.
+	POBJ_TEXT_BENCH,    ///< Bench price text field.
+	POBJ_TEXT_BIN,      ///< Bin   price text field.
+	POBJ_TEXT_LAMP,     ///< Lamp  price text field.
+};
+
+/**
+ * Widget parts of the path objects build GUI.
+ * @ingroup gui_group
+ */
+static const WidgetPart _path_objects_build_gui_parts[] = {
+	Intermediate(0, 1),
+		Intermediate(1, 0),
+			Widget(WT_TITLEBAR, INVALID_WIDGET_INDEX, COL_RANGE_DARK_GREEN), SetData(GUI_PATH_OBJECTS_TITLE, GUI_TITLEBAR_TIP),
+			Widget(WT_CLOSEBOX, INVALID_WIDGET_INDEX, COL_RANGE_DARK_GREEN),
+		EndContainer(),
+		Widget(WT_PANEL, INVALID_WIDGET_INDEX, COL_RANGE_DARK_GREEN),
+			Intermediate(3, 2),
+				Widget(WT_IMAGE_BUTTON, POBJ_BUTTON_BENCH, COL_RANGE_DARK_GREEN), SetData(SPR_GUI_BENCH, GUI_PATH_OBJECTS_BENCH), SetMinimalSize(32, 32),
+				Widget(WT_LEFT_TEXT,    POBJ_TEXT_BENCH,   COL_RANGE_DARK_GREEN), SetData(STR_ARG1,      STR_NULL),
+				Widget(WT_IMAGE_BUTTON, POBJ_BUTTON_BIN,   COL_RANGE_DARK_GREEN), SetData(SPR_GUI_BIN,   GUI_PATH_OBJECTS_BIN),   SetMinimalSize(32, 32),
+				Widget(WT_LEFT_TEXT,    POBJ_TEXT_BIN,     COL_RANGE_DARK_GREEN), SetData(STR_ARG1,      STR_NULL),
+				Widget(WT_IMAGE_BUTTON, POBJ_BUTTON_LAMP,  COL_RANGE_DARK_GREEN), SetData(SPR_GUI_LAMP,  GUI_PATH_OBJECTS_LAMP),  SetMinimalSize(32, 32),
+				Widget(WT_LEFT_TEXT,    POBJ_TEXT_LAMP,    COL_RANGE_DARK_GREEN), SetData(STR_ARG1,      STR_NULL),
+		EndContainer(),
+	EndContainer(),
+};
+
+PathObjectGui::PathObjectGui() : GuiWindow(WC_PATH_OBJECTS, ALL_WINDOWS_OF_TYPE)
+{
+	this->type = nullptr;
+	this->SetupWidgetTree(_path_objects_build_gui_parts, lengthof(_path_objects_build_gui_parts));
+}
+
+PathObjectGui::~PathObjectGui()
+{
+	SetSelector(nullptr);
+}
+
+/**
+ * Sets what kind of path object type is currently being placed.
+ * @param t Type to select.
+ */
+void PathObjectGui::SetType(const PathObjectType *t)
+{
+	if (this->type == t) {
+		this->type = nullptr;
+	} else {
+		this->type = t;
+	}
```

Comment:
Function doesn't do what the description claims, it may end up without selection after a call.

Reply:
Could be a one-liner, although this is fine too.